File: src/main/java/eu.xap3y.xagui/VirtualMenu.java
package eu.xap3y.thepit.api.gui;

import eu.xap3y.thepit.ThePit;
import eu.xap3y.thepit.api.iface.VirtualMenuInterface;
import eu.xap3y.xagui.interfaces.GuiInterface;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;

public abstract class VirtualMenu<T> implements VirtualMenuInterface<T> {
    private final GuiInterface gui;

    public VirtualMenu(GuiInterface gui) {
        this.gui = gui;
    }

    public VirtualMenu(String name, int rows) {
        this.gui = ThePit.xagui.createMenu(name, rows);
    }

    public void open(@NotNull Player player) {
        getGui().open(player);
    }

    public void open(int page, @NotNull Player player) {
        getGui().open(page, player);
    }

    public @NotNull GuiInterface getGui() {
        return this.gui;
    }

    public @NotNull GuiInterface build() {
        return getGui();
    }

    @Override
    public abstract GuiInterface build(@NotNull T o);

    public void open(@NotNull Player player, @NotNull T o) {
        GuiInterface gui = build(o);
        gui.open(player);
    }
}

File: src/main/kotlin/eu/xap3y/xagui/VirtualMenu.kt
package eu.xap3y.xagui

import eu.xap3y.xagui.interfaces.GuiInterface
import eu.xap3y.xagui.interfaces.VirtualMenuInterface
import org.bukkit.entity.Player

abstract class VirtualMenu<T> : VirtualMenuInterface<T> {
    val gui: GuiInterface

    constructor(name: String, rows: Int, instance: XaGui) {
        this.gui = instance.createMenu(name, rows)
    }

    constructor(gui: GuiInterface) {
        this.gui = gui
    }

    fun open(player: Player) {
        gui.open(player)
    }

    fun open(page: Int, player: Player) {
        gui.open(page, player)
    }

    fun build(): GuiInterface {
        return gui
    }

    abstract override fun build(o: T): GuiInterface?

    fun open(player: Player, o: T) {
        val gui = build(o)
        gui?.open(player)
    }
}
